proposals/0292-package-registry-service.md

```diff
@@ -280,38 +386,252 @@ import PackageDescription
+
+The user can associate a package scope with a custom registry
+by passing the `--scope` option.
+
+For example,
+a user might resolve all packages with the package scope `example`
+(such as `example.PriorityQueue`)
+to a private registry.
+
+```terminal
+$ swift package-registry set https://internal.example.com/ --scope example
+$ cat .swiftpm/config/registries.json
+```
+
+```json
+{
+  "registries": { 
+    "example": {
+      "url": "https://internal.example.com"
+    }
+  },
+  "version": 1
+}
+
+```
+
+When a custom registry is associated with a package scope,
+package dependencies with that scope are resolved through the provided URL.
+A custom registry may be associated with one or more scopes,
+but a scope may be associated with only a single registry at a time.
+Scoped custom registries override any unscoped custom registry.
+
+#### Unsetting a custom registry
+
+This proposal also adds a new `swift package-registry unset` subcommand
+to complement the `package-registry set` subcommand.
+
+```manpage
+SYNOPSIS
+	swift package-registry unset <url> [options]
+OPTIONS:
+  --global    Apply settings to all projects for this user
+  --scope     Removes the registry's association to a given scope
+```
+
+Running the `package-registry unset` subcommand
+in the root directory of a package
+updates the `.swiftpm/config` file
+to remove the `default` entry in the top-level `registries` key, if present.
+If a `--scope` option is passed,
+only the entry for the specified scope is removed, if present.
+
+#### Global registry configuration
+
+The user can pass the `--global` option to the `set` or `unset` subcommands
+to update the user-level configuration file located at
+`~/.swiftpm/config/registries.json`.
+
+Any default or scoped registries configured locally in a project directory
+override any values configured globally for the user.
+For example,
+consider the following global and local registry configuration files:
+
+```jsonc
+// Global configuration (~/.swiftpm/config/registries.json)
+{
+  "registries": { 
+    "default": {
+      "url": "https://global.example.com"
+    },
+    "foo": {
+      "url": "https://global.example.com"
+    },
+  },
+  "version": 1
+}
+
+// Local configuration (.swiftpm/config/registries.json)
+{
+  "registries": { 
+    "foo": {
+      "url": "https://local.example.com"
+    }
+  },
+  "version": 1
+}
+
+```
+
+Running the `swift package resolve` command with these configuration files
+resolves packages with the `foo` scope
+using the registry located at "https://local.example.com",
+and all other packages
+using the registry located at "https://global.example.com".
+
+In summary,
+the behavior of `swift package resolve` and related commands
+depends on the following factors,
+in descending order of precedence:
+
+* The package manifest in the current directory (`./Package.swift`)
+* Any existing lock file (`./Package.resolved`)
+* Any local configuration (`./.swiftpm/config`)
```

Comment:
./.swiftpm/config/registries.json

Reply:
Has this change officially been adopted? I know [SE-0291](https://github.com/apple/swift-evolution/blob/main/proposals/0291-package-collections.md) codified `~.swiftpm/config` as a directory, but I wasn't able to find any discussion about an equivalent change for the project / local `./.swiftpm`.

Reply:
right now `./.swiftpm/config` is used for mirrors only. what I am trying to suggest here is that this proposal (0292) will define that the content of `./.swiftpm/config` will be relocated to `./.swiftpm/config/mirrors.json` and the local registries configuration will live in `./.swiftpm/config/registries.json`

Reply:
I think this change out of scope for this proposal. It remains an open question about how we'd migrate `.swiftpm/config` to being a directory while maintaining back backwards compatibility, and I think we'd benefit from discussing it independently of this feature.